x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -187,22 +178,25 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
         } else {
             return organicQuery.toQuery(context);
         }
-
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        if (pinnedIds != null || pinnedDocs != null) {
-            return this;
-        } else if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
             List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
             List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
             if (identifiedPinnedIds == null && identifiedPinnedDocs == null) {
-                return this; // not executed yet
-            } else {
-                return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetId, identifiedPinnedIds, identifiedPinnedDocs, null, null);
-            }
+                return this; // Not executed yet
+            } else if ((identifiedPinnedIds != null && identifiedPinnedIds.isEmpty())
+                && (identifiedPinnedDocs != null && identifiedPinnedDocs.isEmpty())) {
+                    return organicQuery; // Nothing to pin here
+                } else if (identifiedPinnedIds != null && identifiedPinnedIds.isEmpty() == false) {
+                    return new PinnedQueryBuilder(organicQuery, truncateList(identifiedPinnedIds).toArray(new String[0]));
+                } else if (identifiedPinnedDocs != null && identifiedPinnedDocs.isEmpty() == false) {
+                    return new PinnedQueryBuilder(organicQuery, truncateList(identifiedPinnedDocs).toArray(new Item[0]));
+                } else {
+                    return this; // Should never happen
+                }
```

Comment:
I don't see how these lists are ever null, down in the async action:
```
pinnedIdsSetOnce.set(appliedRules.pinnedIds().stream().distinct().toList());
pinnedDocsSetOnce.set(appliedRules.pinnedDocs().stream().distinct().toList());
```
If either were `null`, we would have thrown an NPE on `List.stream()`

So, once they are set, they aren't null. The only time the suppliers will return `null` from what I can tell is when the fetch hasn't finished yet.

Reply:
Yeah that's a good point this is prob. too defensive.

Reply:
You're right, this is annoyingly defensive, I've cleaned it up a bit.